### DISCUSSION TITLE
Server: Typing of query resolvers in different file

### DISCUSSION THREAD

**Discussion by wanschi** [trpc/trpc#4363] [Q&A]
Hello, 

we're really enjoying the DX of tRPC so far. However we want to keep our code base a bit cleaner as our amount of routes is growing. Our idea is to move the query resolvers into another file like this:

```
  getEmail: publicProcedure
    .input(
      z.object({
        invitationCode: z.string(),
      }),
    )
    .query(getEmailHandler),
```

The problem is that we're not able to type the handler (`getEmailHandler`) properly. This is how far we got:

`export const getEmailHandler = async (req: ResolveOptions<any>) => {...}`

There is so much magic happening within tRPC that we're not able to see which value we have to set instead of `any`. You can find the `ResolveOptions` type here in this file `@trpc/server/dist/core/internals/utils`.

This is the type of `ResolveOptions`. Its marked as internal, which means we probably shouldn't even use it.

```
export interface ResolveOptions<TParams extends ProcedureParams> {
  ctx: Simplify<TParams['_ctx_out']>;
  input: TParams['_input_out'] extends UnsetMarker
    ? undefined
    : TParams['_input_out'];
}
```

I hope someone has a better idea than us. I also think this is a really common problem as many people don't want to have their business logic in their router definitions.

Thanks!

**Comment by yamatohagi** [comment]
I am stuck in the same place. Any good solution?

  findMany: protectedProcedure.input(MountainFindManySchema).query(async ({ ctx, input }) => {
    mountainFindManyUseCase({ ctx, input });
    
    
    //usecase
    
    
    
export interface ResolveOptions<TParams extends ProcedureParams> {
  ctx: Simplify<TParams['_ctx_out']>;
  input: TParams['_input_out'] extends UnsetMarker ? undefined : TParams['_input_out'];
}

export const mountainFindManyUseCase = async (params: ResolveOptions) => ({
  
});?????

**Comment by yamatohagi** [comment]
is the solution for now


import { ResolveOptions } from '@trpc/server/dist/core/internals/utils';
import { Context } from 'prisma/context';
import { protectedProcedure, t } from 'prisma/trpc/routers/helpers/createRouter';
import { MountainFindManySchema } from 'server/schemas/findManyMountain.schema';

type MountainFindManySchemaType = (typeof MountainFindManySchema)['_type'];

export const mountainFindManyUseCase = async ({
  input,
  ctx,
}: {
  input: MountainFindManySchemaType;
  ctx: Context;
}) => {
  ctx.prisma.mountain.findMany({

  })
};
